### ISSUE TITLE
If Go Install Package Type is Configured, but Go Is Also Installed by Aqua, will it handle order correctly?

### ISSUE THREAD

**Issue by sheldonhull** [aquaproj/aqua#1679]
I have a configuration that installs Go, but also at the bottom of my registry runs a `go install` package.
Is aqua supposed to handle this order correctly and run package installs first, followed by `go build/go install` type commands?

If not would this be a feature request to ensure I can handle both Go runtime + go install in a single config without having to run multiple tagged sets?

**Comment by suzuki-shunsuke** [comment]
aqua installs packages in parallel, so the order is undetermined.
But when aqua installs a `go_install` package, aqua calls `go install`, then aqua installs Go by `Lazy Install` before running `go install`.
So Go should be installed before running `go install`.

Note that if a symbolic link of Go `$AQUA_ROOT_DIR/bin/go` isn't created aqua doesn't install Go by `Lazy Install`.
`aqua i` creates symbolic links of all packages before installing packages, so if a `go_install` package is installed by `aqua i` there is no problem.
But if a `go_install` package is called and is installed by `Lazy Install`, and a symbolic link of Go isn't created, Go isn't installed.

And be also careful installing a same package in parallel by multiple processes can cause race condition.
Multiple processes work independently so they can read and write files, which can cause race condition.
If `Go` and a `go_install` package are installed in parallel, maybe `Go` is installed by multiple processes in parallel.
aqua has a retry mechanism to solve the race condition as much as possible, but this can't solve the issue perfectly.

**Comment by sheldonhull** [comment]
So the solution is to run aqua install --only-link first then run the install so it will resolve go first? And will that mean a conflict if go is installing already and then a go install command calls it and expects to install it or is the proxy smart enough to know its already in progress? 


Cheers!

**Comment by suzuki-shunsuke** [comment]
There are the following ways to install Go and a `go_install` package `X` in a clean environment.

1. `aqua i` : This probably works but rarely race condition occurs
1. `aqua i -l && X`: This probably works but rarely race condition occurs
1. `aqua i || aqua i`: Even if the first `aqua i` fails due to race condition, the second `aqua i` would succeed because Go is installed by the first `aqua i`. This would work
1. `aqua i -l && go version && X`: Go is installed by Lazy Install. This would work
1. `aqua i -t go && X`: Go is installed first using tag. This would work

**Comment by suzuki-shunsuke** [comment]
> So the solution is to run aqua install --only-link first then run the install so it will resolve go first?

Yes. But if Go is installed by multiple goroutines at the same time the race condition may occur.

> And will that mean a conflict if go is installing already and then a go install command calls it and expects to install it or is the proxy smart enough to know its already in progress?

aqua has no lock mechanism so can't know if go installation is already in progress.
But aqua has a retry mechanism.

**Comment by suzuki-shunsuke** [comment]
I'll consider installing Go before `go_install` packages.

**Comment by sheldonhull** [comment]
I followed your advice and ran the install with the tag '--tags first' before the other installs. However, if Go's situation is not unique and Rust faces a similar issue, a YAML option could be added to allow for custom install orders. The default would be no order, but if a user specifies an install order of 1, 2, 3, those packages would be installed serially before the remaining packages.

Although this would be an additional configuration option, it would only be necessary for those seeking more advanced features.

Instead of you needing to maintain some definition of what constitutes a runtime, this offloads that responsibility onto the user which seems a lot more maintainable. 

Maybe that's a useful idea 💡 😁
